Ensure that random_kruskal_mst doesn't generate the same spanning tree every time

InnovativeInventor opened the pull request:
One of the reasons why we only get constant scores in some chains is because of a different bug, which is caused by `random_kruskal_mst` being re-seeded with the same static seed (`1234`) every time it is called. In some cases, when combined with other strange behavior that needs to be chased down, this results in a totally deterministic step function -- the chain never takes more than one step because every proposal is the same. This issue has been partially addressed in this PR.

sunnymatt reviewed (approved):
Woot woot! Looks beautiful. Thanks for catching this @InnovativeInventor 👍 
![image](https://user-images.githubusercontent.com/5581093/126706104-605b1c46-ecfe-44d6-8f72-db6ab7baa155.png)

gabeschoenbach commented:
Cool, this looks like a necessary and welcome fix! The code looks great, reading through it inspired a few questions that might be good to ask here:
- How should I think about something of type `AbstractRNG`? Is the variable `rng` an integer/seed to a random number generator, or a random number generator itself? It's interesting to me that the same `rng` is passed to distinct non-deterministic functions, for example both `sample_subgraph()` and `random_kruskal_mst()`. I assume this is normal/good, but I'd like to better understand why and how including the variable is helpful.
- Harking back to our conversation about `ReCom` being a "family" of Markov chains, I'm noticing that Python GerryChain can use Wilson's UST algorithm to draw spanning trees, but it looks like Julia only uses random MST. Is there a reason we aren't using Wilson's algorithm in the Julia version, given that (I think) it gets us closer to the "right" target distribution? How much of a priority is it to implement that? More broadly, I'm realizing it's unclear to me exactly how all of our different flavors of ReCom currently differ. Maybe it would be helpful to sketch out a spreadsheet of sorts that details this, if we don't have one already...

My only suggestion would be to add the `rng` argument to the list of arguments in the docstrings (lines 68 and 180 in `balance_edges.jl` and `recom.jl`, respectively). Unless there are stylistic reasons to not include defaulted arguments in the docstring, I think including this would help with folks (like me!) who aren't as familiar with the behavior of these random number generators.

gabeschoenbach reviewed (changes requested):
Cool, this looks like a necessary and welcome fix! The code looks great, reading through it inspired a few questions that might be good to ask here:

How should I think about something of type AbstractRNG? Is the variable rng an integer/seed to a random number generator, or a random number generator itself? It's interesting to me that the same rng is passed to distinct non-deterministic functions, for example both sample_subgraph() and random_kruskal_mst(). I assume this is normal/good, but I'd like to better understand why and how including the variable is helpful.
Harking back to our conversation about ReCom being a "family" of Markov chains, I'm noticing that Python GerryChain can use Wilson's UST algorithm to draw spanning trees, but it looks like Julia only uses random MST. Is there a reason we aren't using Wilson's algorithm in the Julia version, given that (I think) it gets us closer to the "right" target distribution? How much of a priority is it to implement that? More broadly, I'm realizing it's unclear to me exactly how all of our different flavors of ReCom currently differ. Maybe it would be helpful to sketch out a spreadsheet of sorts that details this, if we don't have one already...
My only suggestion would be to add the rng argument to the list of arguments in the docstrings (lines 68 and 180 in balance_edges.jl and recom.jl, respectively). Unless there are stylistic reasons to not include defaulted arguments in the docstring, I think including this would help with folks (like me!) who aren't as familiar with the behavior of these random number generators.

InnovativeInventor commented:
Thanks, Gabe for your feedback! To respond to your first question:
- [`AbstractRNG`](https://docs.julialang.org/en/v1/stdlib/Random/#Random.AbstractRNG) is a type that most (if not all) RNGs in Julia need to implement. It is implemented by both the default RNG in Julia and seeded RNGs like [`MersenneTwister`](https://docs.julialang.org/en/v1/stdlib/Random/#Random.MersenneTwister). `Random.default_rng()` is the default *global* rng that Julia implements (it's a Mersenne Twister), which is why this fix works (the previous RNG was local and reseeded on each call). The user should be able to pass any RNG that implements the AbstractRNG type to `recom()`.
- With regard to your second set of questions, see #92 (not sure if it's ready for review though -- @pjrule?). I don't think it is a super-high priority item, but it'd be great for more spanning tree functions to be tested and added.

I've added documentation explaining the `AbstractRNG` type and a link to point the user to the Julia documentation on RNGs. Furthermore, I've also added more type annotations so it's faster and the user understands what needs to be passed to what (Julia speeds up when you type annotate). Let me know what you think!

gabeschoenbach reviewed (approved):
Looks great!
